data/src/main/java/com/microsoft/azure/kusto/data/auth/endpoints/FastSuffixMatcher.java

```diff
@@ -0,0 +1,94 @@
+package com.microsoft.azure.kusto.data.auth.endpoints;
+
+import com.microsoft.azure.kusto.data.Ensure;
+import com.microsoft.azure.kusto.data.StringUtils;
+
+import java.util.*;
+
+public class FastSuffixMatcher {
+    private final int suffixLength;
+    private final Map<String, List<MatchRule>> m_rules;
+
+    /**
+     * Creates a new matcher with the provided matching rules.
+     *
+     * @param rules - One or more matching rules to apply when Match
+     *              is called
+     * @return FastSuffixMatcher
+     */
+    public static FastSuffixMatcher Create(List<MatchRule> rules) {
+        Ensure.argIsNotNull(rules, "rules");
+        int minRuleLength = rules.stream().min(Comparator.comparing(MatchRule::getSuffixLength))
+                .map(MatchRule::getSuffixLength).orElse(0);
+
+        Ensure.isTrue(minRuleLength > 0 && minRuleLength != Integer.MAX_VALUE, "Cannot have a match rule " +
+                "whose length is zero");
+
+        Map<String, List<MatchRule>> processedRules = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        for (MatchRule rule : rules) {
```

Comment:
You can one-liner this:
`        Map<String, List<MatchRule>> processedRules =  rules.stream().collect(Collectors.groupingBy(rule -> StringUtils.GetStringTail(rule.suffix,
minRuleLength)));`

Reply:
this is not case insensitive tho

Reply:
Fair enough. I would have saved/fetched with toLower, but you can keep as is.